### ISSUE TITLE
Cached streams - version-check?

### ISSUE THREAD

**Issue by warappa** [NEventStore/NEventStore#369]
Hi,

shouldn't `EventStoreRepository` check for the stream-version when opening a stream?

```
if (_streams.TryGetValue(streamId, out stream))
{
    return stream;
}
```

If a _cached_ stream is a **higher** version than requested, it may lead to problems as this is not expected by the caller...

I would propose the following implementation:

```
if (_streams.TryGetValue(streamId, out stream))
{
    if (version == Int32.MaxValue ||
        stream.StreamRevision == version)
        return stream;
}
```

**Comment by thomaseyde** [comment]
Three years later, I was bitten by this. The caching simply doesn't work correctly, at least not in a multithreaded scenario. My projections were consistently one commit behind, no matter how long I waited to run them. The only time it worked as expected was at the first run after startup.

I copied the code, removed the caching code, and everything worked. Except from Save, which create a stream instead of opening it. But that was quick to fix.
